numpy/core/src/umath/override.c

```diff
@@ -109,8 +110,15 @@ normalize___call___args(PyUFuncObject *ufunc, PyObject *args,
             Py_DECREF(obj);
         }
     }
+    /* gufuncs accept either 'axes' or 'axis', but not both */
+    if (nkwds >= 2 && (PyDict_GetItemString(*normal_kwds, "axis") &&
+                       PyDict_GetItemString(*normal_kwds, "axes"))) {
+        PyErr_SetString(PyExc_TypeError,
+                        "cannot specify both 'axis' and 'axes'");
+        return -1;
+    }
```

Comment:
Should `axis` be converted to `axes` here, so that `__array_ufunc__` only needs to support `axes`?

Reply:
In general, I think `__array_ufunc__` should just get whatever is passed in, since in principle at least it may act differently than the normal ufunc. In fact, I am somewhat torn about even including the check here; we do just pass on everything else that the user passes in, even if it would give an error message if no `__array_ufunc__` is present. But then, we do check `sig` and `signature`. But then again, that is a deprecation...

Reply:
Isn't the whole point in these normalization function _not_ to pass on whatever was passed? I'd argue that by not normalizing `axis`, we make it harder for overriders to implement it and `axes` correctly.

The question really is - is your description of how `axis` maps to `axes` an invariant of ufuncs, or just one for the specific case of ndarrays? I think that if a short-hand can mean different things for different subclasses, then it's a dangerous shorthand.

Reply:
As I saw it, the normalization was mostly to ensure all but the inputs were in keyword arguments, so that `__array_ufunc__` implementations didn't have to worry about that; `sig` and `signature` seemed to be there just because `sig` is deprecated.

But I can see your argument as well that one should just translate `axis` because we should not encourage `__array_ufunc__` implementations that can handle `axis` but not `axes`. I'll try to see how it looks if I split out the conversion of `axis` to `axes` as a helper function, as you suggest below.

Reply:
I'm finding myself slightly stuck: I now have a nice `_build_axes_tuple` in `ufunc_object.c`, but to use this in `override.c`, I'd need to include it in a `.h` file -- but in `umath`, those seem to contain only stuff to be exported, which this obviously is not. Most logical might be to start a new `ufunc_parser.c` or `ufunc_helper.c` which contains all these helper functions for parsing, but that feels a bit beyond this PR...

An alternative might be just to remove the `override` commit for now, with the logic that `__array_ufunc__` implementations should do whatever they want; they'll get an error anyway if they pass this back in.

Reply:
It's pretty easy to imagine a library implementing arrays with labeled dimensions that would want to allow specifying axes with names instead of numbers. For example, pandas supports `axis='index')`/`axis='columns'` instead of `axis=0`/`axis=1` (xarray calls this argument `dim` instead of `axis`, but it's basically equivalent).

I suppose the `axis` -> `axes` normalization would be basically equivalent, but let's make sure it works for arbitrary Python objects, not just integers.

Reply:
> but let's make sure it works for arbitrary Python objects, not just integers.

+1 on this. 'axis='rows'` for a ufunc like `(i),(i) -> ()` should end up as `axes=[['rows'], ['rows'], []]`

Reply:
The current normalization does work for arbitrary objects. It just puts `axis` in a single-element tuple, and creates a list with that tuple as entries (or an empty one if an argument has no core dimensions).

Reply:
@mhvk: Can you just put it in `numpy\core\src\umath\ufunc_object.h`? Those functions are not exported either.

Reply:
Yes, putting it in `ufunc_object.h` works (I had mistakenly looked in `ufuncobject.h`, which does have things for export...), but it has a problem that then we need even more parsing...  (I'll move to the main thread, since I think this discussion should not disappear after another change.)

Reply:
Yeah, the fact that both of those filenames exist is super confusing...